host-apis/wasi-0.2.0/host_api.cpp

```diff
@@ -2,29 +2,43 @@
 #include "bindings/bindings.h"
 #include "handles.h"
 
+static std::optional<wasi_clocks_monotonic_clock_own_pollable_t> immediately_ready;
+
+size_t poll_handles(vector<WASIHandle<host_api::Pollable>::Borrowed> handles) {
+  auto list = list_borrow_pollable_t{handles.data(), handles.size()};
+  bindings_list_u32_t result{nullptr, 0};
+  wasi_io_poll_poll(&list, &result);
+  MOZ_ASSERT(result.len > 0);
+  const auto ready_index = result.ptr[0];
+  free(result.ptr);
+  return ready_index;
+}
+
 size_t api::AsyncTask::select(std::vector<AsyncTask *> &tasks) {
   auto count = tasks.size();
-  vector<WASIHandle<host_api::Pollable>::Borrowed> handles;
+  std::vector<WASIHandle<host_api::Pollable>::Borrowed> handles;
 
   for (size_t idx = 0; idx < count; ++idx) {
     auto *task = tasks.at(idx);
     auto id = task->id();
 
     if (id == IMMEDIATE_TASK_HANDLE) {
+      if (handles.size() > 0) {
+        if (!immediately_ready) {
+          immediately_ready = wasi_clocks_monotonic_clock_subscribe_duration(0);
+        }
+        handles.emplace_back(immediately_ready.value().__handle);
+        size_t ready_index = poll_handles(std::move(handles));
+        if (ready_index <= idx) {
+          return ready_index;
+        }
```

Comment:
I just realized that this is subtly wrong, my apologies. Instead of comparing to `idx` here, we need to check if `ready_index` is lower than `handles.size() - 1`. I.e., whether the first handle that's ready is any other than `immediately_ready`. And for that to work, we need to store `handles.size()` before the call to `poll_handles`, because we lose ownership of the vector at that point.